### DISCUSSION TITLE
Trace Init duration with FastAPI

### DISCUSSION THREAD

**Discussion by CGarces** [awslabs/aws-lambda-web-adapter#473] [General]
Hi!

This is more a lambda + fastAPI question than aws-lambda-web-adapter itself but...

After read #283 my expectation was fast cold start below 2 seconds.  I have a lambda with a cold start between  2500 ms and 2800 ms

A typical cold start, when point to /api/account in order to check id the user is authenticated.

> REPORT RequestId: 0b863c3f-9f47-4486-b248-2b0366d08cda	Duration: 2.47 ms	Billed Duration: 3 ms	Memory Size: 512 MB	Max Memory Used: 141 MB	Init Duration: 2735.78 ms

Just 3ms to handle the user data but 2735.78 ms to start the lambda.

The same request on warm lambda.

> REPORT RequestId: d93c662f-047a-4c8a-863c-1d043d58837f	Duration: 2.43 ms	Billed Duration: 3 ms	Memory Size: 512 MB	Max Memory Used: 142 MB

	

How can I trace what is goin on during the 2.7 seconds of the cold start?
Any hint to optimize fastAPI startup without provisioned concurrency?

Simplified python code using .zip package.

```python 
from fastapi import Depends, FastAPI, status
from fastapi.responses import JSONResponse
from common import is_user_login, oauth_manager
import boto3
import fastapi
import api
import os


app = FastAPI(docs_url="/api/docs/", redoc_url=None)
oauth_manager.attach_to_app(app)
# Objeto que controla la conexión del usuario
app.include_router(api.router)


client = boto3.client("s3")


@app.get("/api/account")
async def account(
    user_info=Depends(
        oauth_manager.get_user_info_from(["header", "cookie"], auto_error=False)
    ),
    isLogged=Depends(is_user_login),
):
    if user_info:
        return JSONResponse(...)


@app.get("/management/info")
async def info():
    return JSONResponse(...)


@app.post("/uploadurl")
async def get_upload_url(name: str = fastapi.Body(...)):
     url = client.generate_presigned_post(...)

    return url

```

**Comment by bnusunny** [comment]
You could add `-X importtime` option to python startup command. This will log how much time is spent on importing each package.

**Comment by CGarces** [comment]
> import time: self [us] | cumulative | imported package
> import time:      1145 |    1002282 | fastapi
> import time:  206330 |    1080114 | common
> 

I don't know the scale of the numbers this microseconds?

1 second for fastapi, 1 second for my onw logic...
I have fixed the problem implementing 304 responses and "stale-while-revalidate" cache control for cloudfront.

**Comment by bnusunny** [comment]
Yes, the unit is microseconds.
